compiler.rb

```diff
@@ -122,6 +122,13 @@
 
   provider_config = \
     Provider::Config.parse(provider_yaml_path, product_api, version)
+
+  # TODO(alexstephen): Remove when old overrides are deprecated.
+  if provider_config.is_a?(Array)
```

Comment:
This is confusing! `provider_config` can magically be one thing or another. Can we replace the first occurrence of `provider_config` with something like `product_config_or_product_config_and_api` and then explicitly choose one or the other?